### ISSUE TITLE
Free on-device OCR 

### ISSUE THREAD

**Issue by alkoforumm** [wbaumann/SmartReceiptsLibrary#70]
I dont wanna pay anything for the ocr func so any intention to use some ocr framework on the device itself ? Like tensorflow or so..I can live with the processing time and battery los on a mobile device.

**Comment by wbaumann** [comment]
There's no short term intention, but I've been following [TensorFlow Lite](https://www.tensorflow.org/mobile/tflite/). It's still relatively new, so I have done any significant testing on it, but I would eventually like to support local support for privacy/performance reasons. 

With that said, it's probably not something that we'll plan to work on over the course of this year, but it's definitely something that would be nice for the future. If you're interested, you're also welcome to contribute support for this. Thanks!

**Comment by davispuh** [comment]
@wbaumann do you know any good open source OCR library I could use? Basically my idea is that I don't want to use any third-party OCR API but instead self-host it on my own server.

**Comment by davispuh** [comment]
Looks like [Tesseract](https://github.com/tesseract-ocr/tesseract) is way to go.

**Comment by wbaumann** [comment]
Hey @davispuh,

TL;DR is that Tesseract is really good. I've spent a long while digging into the challenges around OCR for receipts, and here's what I've generally found:

* OCR is a reasonably "solved" problem in that there are lots of great open source libraries our there that convert the images to text (e.g. tesseract)
* The challenge for us is to translate text into the appropriate fields (e.g. how do you determine if the name that you scanned is actually the correct merchant or which of the numbers that tesseract finds is the total vs subtotal vs tax vs line item). To handle the identification/categorization of the appropriate fields, we use a variety of machine-learning tactics via https://www.taggun.io/ to convert the text into usable information (eg price, name, category). At best, however, this process is about 85% accurate and costs ~$0.068 per scan. I've done a crazy amount of investigation into the space and seen two core ideologies:
 * Rather than even bothering with the computer vision, you outsource the "scanning" to a human, who will transcribe everything behind the scenes. My main concern (even ignoring privacy worries like https://arstechnica.com/information-technology/2017/11/expensify-acknowledges-potential-privacy-problem-by-calling-it-a-feature/) is that it's only about 97% accurate and can take up to 24 hours to get a result.
 * Computer vision versions are currently less accurate ~80-85%, but they're much faster. I opted for this, so you can immediately review the results and correct anything that is wrong

I don't have the time at the moment to really explore a tesseract-based solution, but let me know if you do any work with it! I'd be super happy to merge it back upstream

**Comment by e-t-l** [comment]
@wbaumann In your research, did you look at [Aspose OCR Receipt Scanner](https://products.aspose.app/ocr/scan-receipt)? It's open source, and making calls to their API is cheaper than Taggun's. There's also an option to self-host the OCR processing in a Docker image, so I could see Smart Receipts using that in two ways:
1. Users can self-host their own OCR servers and do unlimited scans for free
2. Smart Receipts could host a centralized server. You could charge per-scan, or as a potentially more lucrative and popular model, go freemium: free users can do a limited number of scans per month, or they can subscribe to paid plans to get more (at varying price tiers).

Thoughts?

**Comment by banezaka** [comment]
> @wbaumann In your research, did you look at [Aspose OCR Receipt Scanner](https://products.aspose.app/ocr/scan-receipt)? It's open source, and making calls to their API is cheaper than Taggun's. There's also an option to self-host the OCR processing in a Docker image, so I could see Smart Receipts using that in two ways:
> 
> 1. Users can self-host their own OCR servers and do unlimited scans for free
> 2. Smart Receipts could host a centralized server. You could charge per-scan, or as a potentially more lucrative and popular model, go freemium: users can do a limited number of scans per month for free, or they can subscribe to paid plans to get more (at varying price tiers).
> 
> Thoughts?

Gave Aspose OCR a quick glance, but I couldn't find the source code for the actual software, only some SDKs , there were several mentions about free trials and buying the product so curious to know in what way it is open source?
